Question 1

Given an integer `n`, return *`true` if it is a power of two. Otherwise, return `false`*.

An integer `n` is a power of two, if there exists an integer `x` such that `n == 2x`.
**Example 1:**
Input: n = 1 

Output: true

**Example 2:**
Input: n = 16 


Output: true

**Example 3:**
Input: n = 3 

Output: false

In [1]:
def isPowerof2(n:int):
    if n%2 !=0:
        return False
    if n/2==1:
        return True
    else:
        return((isPowerof2(n/2)))
    
    
#driver code
n = 32
isPowerof2(n)    

True

<aside>
💡 **Question 2**

Given a number n, find the sum of the first natural numbers.

**Example 1:**

Input: n = 3 

Output: 6

**Example 2:**

Input  : 5 

Output : 15

</aside>

In [2]:
def recur_sum(n):
    if n <= 1:
        return n
    else:
        return n + recur_sum(n-1)
# driver code

recur_sum(5)
    

15

<aside>
💡 **Question 3**

Given a positive integer, N. Find the factorial of N. 

</aside>

In [3]:
def recur_factorial(n:int):
    if n==1:
        return n
    else:
        return(n*recur_factorial(n-1))

    # driver code
n = 5    
recur_factorial(n)   

120

<aside>
💡 **Question 4**

Given a number N and a power P, the task is to find the exponent of this number raised to the given power, i.e. N^P.

**Example 1 :** 

Input: N = 5, P = 2

Output: 25

**Example 2 :**
Input: N = 2, P = 5

Output: 32

</aside>

In [4]:
def recur_power(N:int,P:int):
    if P==0:
        return 1
    
    return (N*recur_power(N,P-1))

# driver_code
recur_power(2,5)

32

<aside>
💡 **Question 5**

Given an array of integers **arr**, the task is to find maximum element of that array using recursion.

**Example 1:**

Input: arr = {1, 4, 3, -5, -4, 8, 6};
Output: 8

**Example 2:**

Input: arr = {1, 4, 45, 6, 10, -8};
Output: 45

</aside>

In [5]:
def recur_max(arr:list[int]):
    n = len(arr)
    if n==1:
        return arr[0]
    return max(arr[n-1],recur_max(arr[0:n-1]))

# driver code

arr = [1, 4, 25, -5, -4, 8, 6]
recur_max(arr)

25

<aside>
💡 **Question 6**

Given first term (a), common difference (d) and a integer N of the Arithmetic Progression series, the task is to find Nth term of the series.

**Example 1:**

Input : a = 2 d = 1 N = 5
Output : 6
The 5th term of the series is : 6

**Example 2:**

Input : a = 5 d = 2 N = 10
Output : 23
The 10th term of the series is : 23

</aside>

In [6]:
def arithProgression(a:int,d:int,N:int):
    if N==1:
        return a
    a = a+d
    return (arithProgression(a,d,N-1))

# driver code
a = 5
d = 2
N = 10
arithProgression(a,d,N)

23

<aside>
💡 **Question 7**

Given a string S, the task is to write a program to print all permutations of a given string.

**Example 1:**

***Input:***

*S = “ABC”*

***Output:***

*“ABC”, “ACB”, “BAC”, “BCA”, “CBA”, “CAB”*

**Example 2:**

***Input:***

*S = “XY”*

***Output:***

*“XY”, “YX”*

</aside>

In [7]:
def permutations(remaining, candidate=''):
 
    if len(remaining) == 0:
        print(candidate)
 
    for i in range(len(remaining)):
 
        newCandidate = candidate + remaining[i]
        newRemaining = remaining[0:i] + remaining[i+1:]
 
        permutations(newRemaining, newCandidate)
    
# driver code
S = 'ABC'

permutations(S)

ABC
ACB
BAC
BCA
CAB
CBA


<aside>
💡 **Question 8**

Given an array, find a product of all array elements.

**Example 1:**

Input  : arr[] = {1, 2, 3, 4, 5}
Output : 120
**Example 2:**

Input  : arr[] = {1, 6, 3}
Output : 18

</aside>

In [8]:
def recurProduct(arr:list[int]):
    n = len(arr)
    if n==1:
        return arr[0]
    return(arr[n-1]*recurProduct(arr[0:n-1]))

# driver code
arr= [1, 2, 3, 4, 10]
recurProduct(arr)

240